File: backend/app/api/mappings.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.mapping import (
    ApplyMappingRequest,
    ApplyMappingResponse,
    MappingTemplateResponse,
    MappingTemplateSaveRequest,
    ScopedApplyMappingRequest,
    ScopedApplyMappingResponse,
    SourceColumnsResponse,
    SuggestedMappingResponse,
)
from app.services.mapping import (
    MappingError,
    apply_mapping_to_batch,
    apply_mapping_with_scope,
    get_mapping_template,
    get_suggested_mapping_result,
    get_suggested_mapping_result_for_batch,
    infer_source_columns,
    infer_source_columns_for_ticket_type,
    mapping_rows_to_field_mapping,
    normalize_ticket_type_value,
    save_mapping_template,
)

router = APIRouter(prefix="/mappings", tags=["mappings"])
DbSession = Annotated[Session, Depends(get_db)]


@router.get("/batches/{upload_batch_id}/source-columns", response_model=SourceColumnsResponse)
def get_source_columns(upload_batch_id: UUID, db: DbSession) -> SourceColumnsResponse:
    try:
        source_columns = infer_source_columns(db, upload_batch_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

    return SourceColumnsResponse(upload_batch_id=upload_batch_id, source_columns=source_columns)


@router.get("/source-columns", response_model=SourceColumnsResponse)
def get_ticket_type_source_columns(
    db: DbSession,
    project_id: Annotated[UUID, Query(...)],
    ticket_type: Annotated[str, Query(min_length=1, max_length=40)],
    upload_batch_id: UUID | None = None,
) -> SourceColumnsResponse:
    try:
        if upload_batch_id is not None:
            source_columns = infer_source_columns(db, upload_batch_id)
        else:
            source_columns = infer_source_columns_for_ticket_type(db, project_id, ticket_type)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

    return SourceColumnsResponse(
        upload_batch_id=upload_batch_id,
        project_id=project_id,
        ticket_type=normalize_ticket_type_value(ticket_type),
        source_columns=source_columns,
    )


@router.get("/batches/{upload_batch_id}/suggested-mapping", response_model=SuggestedMappingResponse)
def get_suggested_mapping(upload_batch_id: UUID, db: DbSession) -> SuggestedMappingResponse:
    try:
        suggested_mapping = get_suggested_mapping_result_for_batch(db, upload_batch_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except MappingError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    return SuggestedMappingResponse(
        upload_batch_id=upload_batch_id,
        project_id=suggested_mapping.project_id,
        ticket_type=suggested_mapping.ticket_type,
        mapping_source=suggested_mapping.mapping_source,
        mapping=suggested_mapping.mapping,
        source_columns=suggested_mapping.source_columns,
        suggested_mapping=suggested_mapping.mapping,
    )


@router.get("/suggested-mapping", response_model=SuggestedMappingResponse)
def get_project_ticket_type_suggested_mapping(
    db: DbSession,
    project_id: Annotated[UUID, Query(...)],
    ticket_type: Annotated[str, Query(min_length=1, max_length=40)],
    upload_batch_id: UUID | None = None,
) -> SuggestedMappingResponse:
    try:
        suggested_mapping = get_suggested_mapping_result(
            db=db,
            project_id=project_id,
            ticket_type=ticket_type,
            upload_batch_id=upload_batch_id,
        )
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except MappingError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    return SuggestedMappingResponse(
        upload_batch_id=upload_batch_id,
        project_id=suggested_mapping.project_id,
        ticket_type=suggested_mapping.ticket_type,
        mapping_source=suggested_mapping.mapping_source,
        mapping=suggested_mapping.mapping,
        source_columns=suggested_mapping.source_columns,
        suggested_mapping=suggested_mapping.mapping,
    )


@router.post(
    "/templates",
    response_model=MappingTemplateResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_mapping_template(
    request: MappingTemplateSaveRequest,
    db: DbSession,
) -> MappingTemplateResponse:
    try:
        saved_rows = save_mapping_template(
            db=db,
            project_id=request.project_id,
            ticket_type=request.ticket_type,
            mapping=request.mapping,
            notes=request.notes,
        )
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except MappingError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    ticket_type = request.ticket_type.strip().upper()
    return MappingTemplateResponse(
        project_id=request.project_id,
        ticket_type=ticket_type,
        mapping=mapping_rows_to_field_mapping(saved_rows),
        columns=saved_rows,
    )


@router.get("/templates", response_model=MappingTemplateResponse)
def read_mapping_template(
    db: DbSession,
    project_id: Annotated[UUID, Query(...)],
    ticket_type: Annotated[str, Query(min_length=1, max_length=40)],
) -> MappingTemplateResponse:
    saved_rows = get_mapping_template(db, project_id, ticket_type)
    return MappingTemplateResponse(
        project_id=project_id,
        ticket_type=ticket_type.strip().upper(),
        mapping=mapping_rows_to_field_mapping(saved_rows),
        columns=saved_rows,
    )


@router.post("/batches/{upload_batch_id}/apply", response_model=ApplyMappingResponse)
def apply_mapping(
    upload_batch_id: UUID,
    request: ApplyMappingRequest,
    db: DbSession,
) -> ApplyMappingResponse:
    try:
        if request.save_as_default_for_ticket_type:
            suggested_mapping = get_suggested_mapping_result_for_batch(db, upload_batch_id)
            save_mapping_template(
                db=db,
                project_id=suggested_mapping.project_id,
                ticket_type=suggested_mapping.ticket_type,
                mapping=request.mapping or suggested_mapping.mapping,
            )
        result = apply_mapping_to_batch(
            db=db,
            upload_batch_id=upload_batch_id,
            mapping=request.mapping,
            delete_existing=request.delete_existing,
        )
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except MappingError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=str(exc),
        ) from exc

    return ApplyMappingResponse(
        upload_batch_id=result.upload_batch_id,
        status=result.status,
        total_raw_rows=result.total_raw_rows,
        normalized_ticket_count=result.normalized_ticket_count,
        out_of_scope_ticket_count=result.out_of_scope_ticket_count,
        blank_assignment_group_count=result.blank_assignment_group_count,
        assignment_group_not_in_inventory_count=(
            result.assignment_group_not_in_inventory_count
        ),
        duplicate_skipped_count=result.duplicate_skipped_count,
        failed_row_count=result.failed_row_count,
        warnings=result.warnings,
        errors=result.errors,
    )


@router.post("/apply", response_model=ScopedApplyMappingResponse)
def apply_mapping_for_scope(
    request: ScopedApplyMappingRequest,
    db: DbSession,
) -> ScopedApplyMappingResponse:
    try:
        result = apply_mapping_with_scope(
            db=db,
            project_id=request.project_id,
            ticket_type=request.ticket_type,
            upload_batch_id=request.upload_batch_id,
            scope=request.scope,
            mapping=request.mapping,
            delete_existing=request.delete_existing,
            save_as_default_for_ticket_type=request.save_as_default_for_ticket_type,
        )
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except MappingError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=str(exc),
        ) from exc

    return ScopedApplyMappingResponse(
        scope=result.scope,
        project_id=result.project_id,
        ticket_type=result.ticket_type,
        mapping_source=result.mapping_source,
        saved_as_default_for_ticket_type=result.saved_as_default_for_ticket_type,
        batch_results=result.batch_results,
        total_raw_rows=result.total_raw_rows,
        normalized_ticket_count=result.normalized_ticket_count,
        out_of_scope_ticket_count=result.out_of_scope_ticket_count,
        blank_assignment_group_count=result.blank_assignment_group_count,
        assignment_group_not_in_inventory_count=(
            result.assignment_group_not_in_inventory_count
        ),
        duplicate_skipped_count=result.duplicate_skipped_count,
        failed_row_count=result.failed_row_count,
        warnings=result.warnings,
        errors=result.errors,
    )

File: frontend/src/api/mappings.ts
import { requestJson } from "./client";

export type SourceColumn = {
  name: string;
  normalized_name: string;
  occurrence_count: number;
};

export type SourceColumnsResponse = {
  upload_batch_id: string | null;
  project_id: string | null;
  ticket_type: string | null;
  source_columns: SourceColumn[];
};

export type MappingSource = "SAVED_TEMPLATE" | "BUILT_IN_SUGGESTION";
export type ApplyMappingSource = MappingSource | "REQUEST_BODY";
export type ApplyScope = "BATCH" | "TICKET_TYPE";

export type SuggestedMappingResponse = {
  upload_batch_id: string | null;
  project_id: string | null;
  ticket_type: string | null;
  mapping_source: MappingSource;
  mapping: Record<string, string>;
  source_columns: string[];
  suggested_mapping: Record<string, string>;
};

export type MappingTemplateResponse = {
  project_id: string;
  ticket_type: string;
  mapping: Record<string, string>;
};

export type NormalizationErrorSample = {
  row_number: number;
  raw_row_id: string;
  message: string;
};

export type ApplyMappingResponse = {
  upload_batch_id: string;
  status: string | null;
  total_raw_rows: number;
  normalized_ticket_count: number;
  out_of_scope_ticket_count: number;
  blank_assignment_group_count: number;
  assignment_group_not_in_inventory_count: number;
  duplicate_skipped_count: number;
  failed_row_count: number;
  warnings: string[];
  errors: NormalizationErrorSample[];
};

export type BatchApplyMappingResult = ApplyMappingResponse & {
  batch_name: string;
};

export type ScopedApplyMappingResponse = {
  scope: ApplyScope;
  project_id: string;
  ticket_type: string;
  mapping_source: ApplyMappingSource;
  saved_as_default_for_ticket_type: boolean;
  batch_results: BatchApplyMappingResult[];
  total_raw_rows: number;
  normalized_ticket_count: number;
  out_of_scope_ticket_count: number;
  blank_assignment_group_count: number;
  assignment_group_not_in_inventory_count: number;
  duplicate_skipped_count: number;
  failed_row_count: number;
  warnings: string[];
  errors: NormalizationErrorSample[];
};

export type SaveMappingTemplateInput = {
  projectId: string;
  ticketType: string;
  mapping: Record<string, string>;
};

export type ApplyMappingInput = {
  uploadBatchId: string;
  mapping: Record<string, string>;
  deleteExisting: boolean;
  saveAsDefaultForTicketType?: boolean;
};

export type ScopedApplyMappingInput = {
  projectId: string;
  ticketType: string;
  uploadBatchId?: string;
  scope: ApplyScope;
  mapping: Record<string, string>;
  deleteExisting: boolean;
  saveAsDefaultForTicketType: boolean;
};

export function getSourceColumns(uploadBatchId: string): Promise<SourceColumnsResponse> {
  return requestJson<SourceColumnsResponse>(
    `/mappings/batches/${uploadBatchId}/source-columns`
  );
}

export function getSourceColumnsForTicketType(
  projectId: string,
  ticketType: string,
  uploadBatchId?: string
): Promise<SourceColumnsResponse> {
  const query = new URLSearchParams({
    project_id: projectId,
    ticket_type: ticketType,
  });
  if (uploadBatchId) {
    query.set("upload_batch_id", uploadBatchId);
  }
  return requestJson<SourceColumnsResponse>(`/mappings/source-columns?${query.toString()}`);
}

export function getSuggestedMapping(uploadBatchId: string): Promise<SuggestedMappingResponse> {
  return requestJson<SuggestedMappingResponse>(
    `/mappings/batches/${uploadBatchId}/suggested-mapping`
  );
}

export function getSuggestedMappingForTicketType(
  projectId: string,
  ticketType: string,
  uploadBatchId?: string
): Promise<SuggestedMappingResponse> {
  const query = new URLSearchParams({
    project_id: projectId,
    ticket_type: ticketType,
  });
  if (uploadBatchId) {
    query.set("upload_batch_id", uploadBatchId);
  }
  return requestJson<SuggestedMappingResponse>(`/mappings/suggested-mapping?${query.toString()}`);
}

export function saveMappingTemplate(
  input: SaveMappingTemplateInput
): Promise<MappingTemplateResponse> {
  return requestJson<MappingTemplateResponse>("/mappings/templates", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      project_id: input.projectId,
      ticket_type: input.ticketType,
      mapping: input.mapping,
    }),
  });
}

export function getMappingTemplate(
  projectId: string,
  ticketType: string
): Promise<MappingTemplateResponse> {
  const query = new URLSearchParams({
    project_id: projectId,
    ticket_type: ticketType,
  });
  return requestJson<MappingTemplateResponse>(`/mappings/templates?${query.toString()}`);
}

export function applyMapping(input: ApplyMappingInput): Promise<ApplyMappingResponse> {
  return requestJson<ApplyMappingResponse>(
    `/mappings/batches/${input.uploadBatchId}/apply`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        mapping: input.mapping,
        delete_existing: input.deleteExisting,
        save_as_default_for_ticket_type: input.saveAsDefaultForTicketType ?? false,
      }),
    }
  );
}

export function applyMappingForScope(
  input: ScopedApplyMappingInput
): Promise<ScopedApplyMappingResponse> {
  return requestJson<ScopedApplyMappingResponse>("/mappings/apply", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      project_id: input.projectId,
      ticket_type: input.ticketType,
      upload_batch_id: input.uploadBatchId || null,
      scope: input.scope,
      mapping: input.mapping,
      delete_existing: input.deleteExisting,
      save_as_default_for_ticket_type: input.saveAsDefaultForTicketType,
    }),
  });
}
